Return FieldArrayPositions along with Fields in DocumentMatch

viswajithiii opened the pull request:
We use Bleve internally. For our use case, when we received fields from a search result, we also needed to know exactly which element had which field value (for documents with nested arrays in them). 

This patch returns `FieldArrayPositions` in the DocumentMatch, which basically is a companion array to `Fields` which returns the array positions of the corresponding element. We did it this way so that we didn't break any clients of the existing struct. We've been using this patch internally for a while now, and it's worked for us, so we're sending this PR out in case you folks upstream think it would be useful.

viswajithiii commented:
@abhinavdangeti What do you think of this change?

abhinavdangeti commented on search/search.go:
```diff
@@ -97,9 +97,17 @@ type DocumentMatch struct {
 	// Fields contains the values for document fields listed in
 	// SearchRequest.Fields. Text fields are returned as strings, numeric
-	// fields as float64s and date fields as time.RFC3339 formatted strings.
+	// fields as float64s, boolean fiels as bools, and date fields as
```
typo: fields

abhinavdangeti commented on search/search.go:
```diff
@@ -97,9 +97,17 @@ type DocumentMatch struct {
 
+	// FieldArrayPositions is a companion map to Fields which returns the array positions
+	// corresponding to each element in Fields.
+	FieldArrayPositions map[string][]ArrayPositions
```
\`json:"fields,omitempty"\`

viswajithiii commented:
Yeah, it wasn't possible to get the actual field values from the `FieldTermLocationMap`. (The term was a lowercase string for string fields, and an opaque byte sequence for non-string fields.)

I've addressed your comments.

abhinavdangeti commented:
@viswajithiii Before we get this merged, can you to sign the CLA, as described here: https://github.com/blevesearch/bleve/blob/master/CONTRIBUTING.md

viswajithiii commented:
@abhinavdangeti I think I already did that a couple days ago. Can you confirm that I did things correctly?
![image](https://user-images.githubusercontent.com/3502044/47750289-4c44c280-dc4c-11e8-98d0-926c4b5d94dc.png)

mschoch commented:
The original intention was that the returned sliced should recreate the structure of the original slice.  I can see that the current implementation probably doesn't always do that. But that would be considerably more elegant than this solution.

Right now people already complain that the returned stored fields don't resemble the original values (numbers and dates are returned in an internal representation).  Returning the slice position info as another field seems like its taking us further down this path.  If this patch returns a slice of array positions so that the original can be recreated, why not just do this recreation inside of bleve?

viswajithiii commented:
Closing since stale.
